### ISSUE TITLE
Display Module

### ISSUE THREAD

**Issue by myjeeed** [myjeeed/MakerWear#87]
Now that we are exploring programming tools, another useful module to work on is the display module.
This tiny display looks very promising: https://www.sparkfun.com/products/12923 but it's too big (26mm x 26mm x 16mm) - take a look at the following photo to compare their sizes:
![image](https://cloud.githubusercontent.com/assets/9346800/19086454/33d2d048-8a3c-11e6-8d95-442151e43dd2.png)

Unfortunately with the current architecture, we are not able to have this display on anything else other than the programmable module it self. However, this allows us to show the following things on the display:
1. Text
2. Numbers
3. Shapes and Icons
4. Sprite characters
5. and user-defined stuff

I think that this is also extremely important for the high-ceiling design principle that we have in MakerWear's visual programming tool.

Please let me know what you think about this.
Also it would be great if you could let us know about any other display components that we could use. There is also a custom built display option which is kinda difficult, but would allow us to have a hexagonal shape display!!

**Comment by myjeeed** [comment]
We could build a custom display using SMD 1.6x1.6mm LEDs which requires ~20 digital I/O pins for controlling, which means that it has to have its own ATmega328 controller that should also communicate with the module's microcontroller! 
![image](https://cloud.githubusercontent.com/assets/9346800/19087067/b6b215a8-8a3e-11e6-8777-271cba94c467.png)

So I think we should definitely try to find a tiny display that is available in the market... the same thing that we did for the Servo motors!

**Comment by myjeeed** [comment]
This LCD looks promising:
https://www.sparkfun.com/products/13003
but still too big... 20.32mm x 20.32mm
![image](https://cloud.githubusercontent.com/assets/9346800/19089478/6eab6980-8a48-11e6-9192-d2a3000693f1.png)
![image](https://cloud.githubusercontent.com/assets/9346800/19089577/e6bb5a8e-8a48-11e6-9721-7630d61256c3.png)

**Comment by myjeeed** [comment]
Look what I've found: https://www.seeedstudio.com/0.5-Inch-OLED-display-Arduino-shield-p-2194.html
still, a little bit too big, but we might be able to cut those excessive parts or just hide them under the top layer cover.
![image](https://cloud.githubusercontent.com/assets/9346800/19090084/efce67ea-8a4a-11e6-94da-23e920a7ab5f.png)

![image](https://cloud.githubusercontent.com/assets/9346800/19089699/72af5536-8a49-11e6-96fa-14d6f824b9b6.png)

**Comment by jmcpeak2** [comment]
I was able to get the smaller of the two displays working (the LD7032)

I found the pinout of the PCB from this document (www.seeedstudio.com/document/pdf/OCELL_use_guide2.pdf) and then wired those pins according to the Arduino pins shown for the LD7032 in the hello_world example from the library (https://github.com/olikraus/u8glib)

I will now experiment with changing the size of the characters, or perhaps enabling character scrolling so we can fit more than a few letters on the screen

**Comment by jonfroehlich** [comment]
I want to see a picture or video! :)

**Comment by jmcpeak2** [comment]
Here is a video showing a basic animation. This was very easy to make using the library
https://drive.google.com/open?id=0BxXWEwzrNM9ZMllOUmtrVzlYOUU

This one is the same example, but without any delay to demonstrate the maximum animation speed possible
https://drive.google.com/open?id=0BxXWEwzrNM9ZYks2c2ctRWo2VkE

This example shows several different font sizes
https://drive.google.com/open?id=0BxXWEwzrNM9ZNmRBd3RIZ2dndTA

There are a variety of different fonts included with the library, listed [here](https://github.com/olikraus/u8glib/wiki/fontgroup). Each of the "fontgroups" listed has one font in a variety of sizes. However, I've noticed that using multiple fonts takes up a lot of memory on the Arduino. When I tried to do more than 5 in a single sketch, I ran out of memory.

I think there may be a way to just include a single font, and then scale the screen to change its size. I will need to look into this.

**Comment by jonfroehlich** [comment]
This is great progress. Thanks @jmcpeak2. Do you guys have some sketches for some initial visualizations/UIs we would like to support?

**Comment by jmcpeak2** [comment]
I think the reliability of the [smaller screens](https://www.seeedstudio.com/0.5-Inch-OLED-display-Arduino-shield-p-2194.html) may be an issue.

One of the two we bought didn't seem to work at all, and the second one stopped working suddenly today. It was freezing and displaying pixels in the wrong location for a little bit, and then stopped working entirely.

I'll try to figure out if I was doing something wrong (and broke the screens), but perhaps we should look for alternatives (such as the larger SparkFun screen)

**Comment by jonfroehlich** [comment]
I agree. 
1. Search online and see if others are experiencing same issue
2. Try contacting manufacturer
3. Use alternative like Sparkfun

Do in parallel

**Comment by myjeeed** [comment]
The problem was with the flat connections from the PCB to the OLED. Unfortunately, it seems that they got bent and the connection is now broken.

However, we got it to work again by gluing the OLED display on the board to make sure that it maintains a solid connection.

**Comment by jonfroehlich** [comment]
How about the Sparkfun alternatives? Have you played with those yet? If so, I'd like to see a video demo.

**Comment by myjeeed** [comment]
Yep, we got it to work. And we also found out that many of those capacitors and resistors are unnecessary (Jason desoldered them, and it was still working).
We will probably use the Sparkfun module as it has a better library, documentation, more reliable and faster to compile. Currently Jason is trying to redesign the schematics with minimal components so that I could design the board layout. We will desolder the OLED from the sparkfun module, and solder it on our own custom-built PCB.

**Comment by jonfroehlich** [comment]
Great! I'm guessing those caps/resistors are there for a reason--e.g., to smooth signal, protect from voltage spikes, etc. Thoughts?

Could probably source the OLED directly. Doesn't Sparkfun documentation state specifically what model it is--then we could check digikey, mouser, etc.

**Comment by myjeeed** [comment]
This is the OLED that is used on the Sparkfun module: ug-6448hlbeg03
However, I wasn't able to find any distributer that we could buy from. (everyone is out of stock!)

**Comment by myjeeed** [comment]
I made a standalone version of the OLED display that moves a ball according to the input voltage of the Module.
Here are some videos of it:
https://www.dropbox.com/s/8ix5nt8q2l4n837/testingOLED_button.mov?dl=0
https://www.dropbox.com/s/a7tfg1cwy79lr4z/testingOLED_distanceSensor.mov?dl=0
https://www.dropbox.com/s/66ug46b2z3stga0/testingOLED_sineWave.mov?dl=0

**Comment by jonfroehlich** [comment]
Yes, this is AAWWEEEEEESSSSSOME! Thanks for the great videos as well. Woohoo! Next generation MakerWear, here we come!

**Comment by myjeeed** [comment]
We decided to have a button on the Programmable module (with the OLED display) to get it into a visualization mode to show the value of all I/O signals + the I/Os that are active.
Something like this:
![oled_display](https://cloud.githubusercontent.com/assets/9346800/20283374/c328f3c0-aa86-11e6-84f2-a5bd449ddae2.png)
